### PR TITLE
Fixed integer overflow when matching PVPVC claims

### LINKED ISSUE #49911: Storage Capacity of persistent volumes in k8s 1.7.2 reduced to 63 bits (signed 64 bits?)
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
Earlier we had setup 12E and NFS storage for PersistentVolume and this size is near 64bit. Now I see that 12E fails for NFS and the upper limit I could get to was 2.3E (61 bits). For local storage, I can get up to 9.2E (63 bits). The PV gets set up but any pvc remains in "Pending" status.

**What you expected to happen**:
I expected behavior to be as earlier, with 12E being able to satisfy claims. If it's not possible to have a 12E, I would expect at least one common size across all PV types. Else, some documentation would be useful.

**How to reproduce it (as minimally and precisely as possible)**:
Examples of local and nfs yaml files are as below:

# Local: 
pvc.yaml (local):
```
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: task-pv-claim
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 1E
```

pv.yaml (local, working):
```
kind: PersistentVolume
apiVersion: v1
metadata:
  name: task-pv-volume
  labels:
    type: local
spec:
  capacity:
    storage: 9.2E
  accessModes:
    - ReadWriteMany
  hostPath:
    path: "/tmp/data"
```

pv.yaml (local, failing):
```
kind: PersistentVolume
apiVersion: v1
metadata:
  name: task-pv-volume
  labels:
    type: local
spec:
  capacity:
    storage: 12E
  accessModes:
    - ReadWriteMany
  hostPath:
    path: "/tmp/data"
```
Working local:
```
kubectl get pv
NAME             CAPACITY   ACCESSMODES   RECLAIMPOLICY   STATUS      CLAIM                    STORAGECLASS   REASON    AGE
task-pv-volume   9200P      RWX           Retain          Available                                                     13s

kubectl get pvc
NAME             STATUS    VOLUME           CAPACITY   ACCESSMODES   STORAGECLASS   AGE
task-pv-claim    Bound     task-pv-volume   9200P      RWX                          5s
```

Failing local:
```
kubectl get pv
NAME             CAPACITY   ACCESSMODES   RECLAIMPOLICY   STATUS      CLAIM                    STORAGECLASS   REASON    AGE
task-pv-volume   12E        RWX           Retain          Available                                                     5m

kubectl get pvc
NAME             CAPACITY   ACCESSMODES   RECLAIMPOLICY   STATUS      CLAIM                    STORAGECLASS   REASON    AGE
task-pv-claim    Pending                                                            1m
```

# NFS:
nfs-pvc.yaml:
```
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: test-claim
  labels:
    app: test
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 1E
  volumeName: test-share
```
nfs-pv.yaml (working, IP redacted):
```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: test-share
  labels:
    app: test
spec:
  capacity:
    storage: 2.3E
  accessModes:
    - ReadWriteMany
  persistentVolumeReclaimPolicy: Retain
  nfs:
    server: xxx.xxx.xxx.xxx
    path: /tmp/test
```

nfs-pv.yaml (failing, IP redacted):
```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: test-share
  labels:
    app: test
spec:
  capacity:
    storage: 12E
  accessModes:
    - ReadWriteMany
  persistentVolumeReclaimPolicy: Retain
  nfs:
    server: xxx.xxx.xxx.xxx
    path: /tmp/test
```

Working NFS:
```
kubectl get pv
NAME         CAPACITY   ACCESSMODES   RECLAIMPOLICY   STATUS    CLAIM                STORAGECLASS   REASON    AGE
test-share   2300P      RWX           Retain          Bound     default/test-claim                            31s

kubectl get pvc
NAME         STATUS    VOLUME       CAPACITY   ACCESSMODES   STORAGECLASS   AGE
test-claim   Bound     test-share   2300P      RWX                          4s
```
Failing NFS:
```
kubectl get pv
NAME         CAPACITY   ACCESSMODES   RECLAIMPOLICY   STATUS    CLAIM                STORAGECLASS   REASON    AGE
test-share   12E      RWX           Retain          Bound     default/test-claim                            4s

kubectl get pvc
NAME         STATUS    VOLUME       CAPACITY   ACCESSMODES   STORAGECLASS   AGE
test-claim   Pending   test-share   0                                       3m
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): Kubernetes 1.7.2
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release): Centos 7.2
- Kernel (e.g. `uname -a`):3.10.0-327.3.1.el7.x86_64 #1 SMP Wed Dec 9 14:09:15 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: rpm
- Others:

### PR DESCRIPTION
**Pull request by davidz627**
Fixes #49911

Fixed integer overflow when matching PVPVC claims. Added test to guard this behavior.

### CODE REVIEW

File: pkg/controller/volume/persistentvolume/index_test.go [hunk 1]
```diff
@@ -129,6 +129,15 @@ func TestMatchVolume(t *testing.T) {
 				pvc.Spec.StorageClassName = &classSilver
 			}),
 		},
+		"successful-match-vlarge": {
```
Comment (msau42) [inline pkg/controller/volume/persistentvolume/index_test.go]: I would also add some more test cases:
- a negative case where the requested size is larger than the PV size.
- positive case where the requested size is equal to the PV size

File: pkg/controller/volume/persistentvolume/framework_test.go [hunk 2]
```diff
@@ -805,6 +805,7 @@ var (
 	classExternal                string = "external"
 	classUnknownInternal         string = "unknown-internal"
 	classUnsupportedMountOptions string = "unsupported-mountoptions"
+	classVLarge                  string = "vlarge"
```
Comment (msau42) [inline pkg/controller/volume/persistentvolume/framework_test.go]: maybe just call it "large" or "very-large"

**Comment by rootfs** [comment]
@davidz627 fix test failures

/assign @jsafrane 
/unassign

**Comment by jsafrane** [comment]
nice find!
/approve no-issue

**Comment by msau42** [comment]
Fixes #49911

**Comment by msau42** [comment]
Try running hack/update-bazel.sh to generate the build file changes

**Comment by jsafrane** [comment]
Thanks for the PR!
/lgtm

**Comment by jsafrane** [comment]
@davidz627, I just noticed there are several commits in this PR. Please squash them together, we don't want to pollute our commit history.

/lgtm cancel

**Pull request merged**